cel/canonical_eventlog_test.go

```diff
@@ -56,7 +59,99 @@ func TestCELEncodingDecoding(t *testing.T) {
 	if len(digestsMap[crypto.SHA1]) != 20 {
 		t.Errorf("SHA1 digest length doesn't match")
 	}
-	if reflect.DeepEqual(decodedcel.Records, (*cel).Records) == false {
+	if !reflect.DeepEqual(decodedcel.Records, cel.Records) {
 		t.Errorf("decoded CEL doesn't equal to the original one")
 	}
 }
+
+func TestCELMeasureAndReplay(t *testing.T) {
+	tpm := test.GetTPM(t)
+	defer client.CheckedClose(t, tpm)
+
+	cel := &CEL{}
+	measuredHashes := []crypto.Hash{crypto.SHA256, crypto.SHA1, crypto.SHA512}
+
+	someEvent := make([]byte, 10)
+	cosEvent := CosTlv{someEvent}
+	someEvent2 := make([]byte, 10)
+
+	rand.Read(someEvent2)
+	cosEvent2 := CosTlv{someEvent2}
+	appendOrFatal(t, cel, tpm, test.DebugPCR, measuredHashes, cosEvent)
+	appendOrFatal(t, cel, tpm, test.DebugPCR, measuredHashes, cosEvent2)
+
+	appendOrFatal(t, cel, tpm, test.ApplicationPCR, measuredHashes, cosEvent2)
+	appendOrFatal(t, cel, tpm, test.ApplicationPCR, measuredHashes, cosEvent)
+	appendOrFatal(t, cel, tpm, test.ApplicationPCR, measuredHashes, cosEvent)
+
+	replay(t, cel, tpm, measuredHashes,
+		[]int{test.DebugPCR, test.ApplicationPCR} /*shouldSucceed=*/, true)
+}
+
+func TestCELReplayFailTampered(t *testing.T) {
+	tpm := test.GetTPM(t)
+	defer client.CheckedClose(t, tpm)
+
+	cel := &CEL{}
+	measuredHashes := []crypto.Hash{crypto.SHA256, crypto.SHA1, crypto.SHA512}
+
+	someEvent := make([]byte, 10)
+	cosEvent := CosTlv{someEvent}
+	someEvent2 := make([]byte, 10)
+
+	rand.Read(someEvent2)
+	cosEvent2 := CosTlv{someEvent2}
+	appendOrFatal(t, cel, tpm, test.DebugPCR, measuredHashes, cosEvent)
+	appendOrFatal(t, cel, tpm, test.DebugPCR, measuredHashes, cosEvent2)
+
+	appendOrFatal(t, cel, tpm, test.ApplicationPCR, measuredHashes, cosEvent2)
+	appendOrFatal(t, cel, tpm, test.ApplicationPCR, measuredHashes, cosEvent)
+	appendOrFatal(t, cel, tpm, test.ApplicationPCR, measuredHashes, cosEvent)
+
+	modifiedRecord := cel.Records[3]
+	for hash := range modifiedRecord.Digests {
+		newDigest := make([]byte, hash.Size())
+		rand.Read(newDigest)
+		modifiedRecord.Digests[hash] = newDigest
+	}
+	replay(t, cel, tpm, measuredHashes,
```

Comment:
We should test replay using a subset/superset of the measuredHashes, to see its behavior

Reply:
Done.